### DISCUSSION TITLE
Handlr no longer opens any files on my mimeapps.list

### DISCUSSION THREAD

**Discussion by charbelnicolas** [Anomalocaridid/handlr-regex#68] [Q&A]
After the latest update handlr no longer opens any file on the mimeapps.list. What is the point of having a "handler" if it doesn't open any files? Am I to set manually all the file types handlr should open by handlr set?

**Comment by charbelnicolas** [comment]
Wow, i used handlr set .png feh.desktop and it wiped my whole curated mimeapps.list with no way of undoing. What a crock of shit.

**Reply by Anomalocaridid** [reply]
I'm very, very sorry that this happened. What version were you using? Could you tell me what you remember you had in your `mimeapps.list`? Any information would be helpful.

I don't blame you if you permanently lost trust in my program, but I'll fix this so no one else has to go through this.

**Reply by Anomalocaridid** [reply]
I have fixed this particular in the [fix-discussion-68](https://github.com/Anomalocaridid/handlr-regex/tree/fix-discussion-68) branch. Would you be able to clone it and test it? It is alright if you cannot.

**Comment by charbelnicolas** [comment]
Just because a mimeapps.list doesn't have a fucking [Added Associations] it stops working... what a piece of shit.

**Reply by Anomalocaridid** [reply]
Again, I am very, very sorry. But this information helps immensely and I appreciate you mentioning it.

**Reply by Anomalocaridid** [reply]
Also are you using v0.11.0? If so, I might know what could be causing this and I'll should be able to fix it soon.

**Reply by charbelnicolas** [reply]
Hello, yes I'm using version 0.11 the package just got updated today in Arch Linux and I wasn't able to open any files as I normally do.

**Comment by charbelnicolas** [comment]
This doesn't work anymore either `x-scheme-handler/terminal=org.codeberg.dnkl.foot.desktop;` I can't open files with neovim.

`text/*=nvim.desktop;`

**Reply by Anomalocaridid** [reply]
Would you mind posting the contents of `org.codeberg.dnkl.foot.desktop`? Does it only fail with neovim or do other terminal programs not work either?

**Reply by charbelnicolas** [reply]
```
[Desktop Entry]
Type=Application
Exec=foot
Icon=foot
Terminal=false
Categories=System;TerminalEmulator;
Keywords=shell;prompt;command;commandline;

Name=Foot
GenericName=Terminal
Comment=A wayland native terminal emulator
StartupWMClass=foot
```

**Reply by Anomalocaridid** [reply]
Thank you!

**Reply by charbelnicolas** [reply]
I don't have other programs that I use on the terminal other than nnn file manager which never opens up if I set `inode/directory=nnn.desktop` mime type so I wouldn't really know, sorry.

**Reply by Anomalocaridid** [reply]
No worries, thanks anyways. I'm not sure what the issue could because I have been able to use `nnn` like that with `handlr`, so it might be worth another try unless it was recent.

**Reply by charbelnicolas** [reply]
> No worries, thanks anyways. I'm not sure what the issue could because I have been able to use `nnn` like that with `handlr`, so it might be worth another try unless it was recent.

I think the nnn issue mostly has to do with chromium when trying to open the download folder. It has never worked.

**Reply by Anomalocaridid** [reply]
Unfortunately, I am having difficulty replicating your issue.

How are you using `handlr`? (e.g. directly from command line, shadowing `xdg-open`, etc.) Does this issue arise when opening the download folder through Chromium, since you mentioned it? If so, does it still occur with other web browsers?

Please feel free to mention any other information regarding this issue that you think is pertinent or helpful.

**Reply by charbelnicolas** [reply]
> Unfortunately, I am having difficulty replicating your issue.
> 
> How are you using `handlr`? (e.g. directly from command line, shadowing `xdg-open`, etc.) Does this issue arise when opening the download folder through Chromium, since you mentioned it? If so, does it still occur with other web browsers?
> 
> Please feel free to mention any other information regarding this issue that you think is pertinent or helpful.

I'm sorry, I'm not an expert when it comes to these issues, if the shadowing isn't done automatically on xdg-open, I'm pretty sure I'm not doing it then. I don't expect chromium to work, it has all kinds of quirks but I just deal with it. On the other hand the nvim issue does affect me considerably since I open a lot of different files in nvim daily and handlr-regex version 0.10.1 works.

**Reply by Anomalocaridid** [reply]
No worries. You're already helping me a lot by cooperating with me on this. Sorry if I kind of overloaded you with questions.

How do you use handlr-regex to open files in nvim? Do you open the file by directly running the `handlr` command?

**Reply by charbelnicolas** [reply]
I added a xdg-open.sh file to $HOME/.local/bin with

```
#!/bin/sh

handlr open "$@"
```

but that didn't seem to help with chromium. Are you telling me that you can open nnn on the terminal if you click on the open download folder in chromium directly?

**Reply by charbelnicolas** [reply]
> No worries. You're already helping me a lot by cooperating with me on this. Sorry if I kind of overloaded you with questions.
> 
> How do you use handlr-regex to open files in nvim? Do you open the file by directly running the `handlr` command?

I open files through nnn file manager. If I open them with version 0.11 I only get the nvim process running without the terminal. If I use version 0.10.1 doing the exact same thing, opening the file through nnn, it opens nvim in the foot terminal along with the file.

**Reply by Anomalocaridid** [reply]
> Are you telling me that you can open nnn on the terminal if you click on the open download folder in chromium directly?

No, I assumed that's what you meant when you said you were using `nnn` with Chromium. I don't use Chromium, so I wouldn't know. Sorry if I confused you.

I was just trying to cover my bases to try to get as much information about your problem as possible because I cannot figure out how to fix it if I cannot get the problem to occur for me.

Also the script should be named xdg-open, not xdg-open.sh.

>I open files through nnn file manager.

Could you walk me through the exact steps you take to open files through nnn?

**Reply by charbelnicolas** [reply]
> > Are you telling me that you can open nnn on the terminal if you click on the open download folder in chromium directly?
> 
> No, I assumed that's what you meant when you said you were using `nnn` with Chromium. I don't use Chromium, so I wouldn't know. Sorry if I confused you.
> 
> I was just trying to cover my bases to try to get as much information about your problem as possible because I cannot figure out how to fix it if I cannot get the problem to occur for me.
> 
> Also the script should be named xdg-open, not xdg-open.sh.
> 
> > I open files through nnn file manager.
> 
> Could you walk me through the exact steps you take to open files through nnn?

I open files on nnn by pressing enter or right arrow when a file is highlighted. Please check the video below:

[https://drive.google.com/file/d/1VVc7-d-qf7NMFc0RxCsa6Wbja9eCxFgx/view?usp=sharing](https://drive.google.com/file/d/1VVc7-d-qf7NMFc0RxCsa6Wbja9eCxFgx/view?usp=sharing)

**Reply by Anomalocaridid** [reply]
Thanks. Sorry to keep asking you questions, but how do you have nnn configured to use handlr to open files? I assume it is not just something that is set up by default, right?

**Reply by charbelnicolas** [reply]
I only have some launch parameters (`-USJHd`), I've never had to set up anything else for it. Is there something I should be doing?

**Reply by Anomalocaridid** [reply]
No, I'm just trying to get as much information as I can on your setup because I am still having difficulty replicating the issue you are having with using `handlr` to open files in `nvim` through `nnn`. Basically, I am trying to get the bug you're experiencing to occur on my system so that I can figure out what is causing it.

**Reply by Anomalocaridid** [reply]
Even using `nnn`, `foot`, and `nvim`, with handlr-regex v0.11.0, it works fine for me and I am not sure why. Could you maybe go back to v0.11.0 just to double-check with the mimeapps.list that is working fine for you with v0.10.1? Of course, please make sure to back up mimeapps.list beforehand just in case.

**Reply by charbelnicolas** [reply]
I already tried version 0.11 today with the `xdg-open` shadowing on `.local/bin` and it did not work either (Using the exact same mimeapps.list). Are you on kde or gnome? What is your DE? I'm on labwc using wayland. I don't know if that makes a difference....

**Reply by Anomalocaridid** [reply]
I use Hyprland, a Wayland window manager. The distro I use is NixOS. It might make a difference, but if so, I do not know what it would be.

Would you mind posting your mimeapps.list?

**Reply by charbelnicolas** [reply]
I set up a shortcut for labwc using `handlr open /home/charbel/.log.txt` and it works, it opens the foot terminal, nvim and the .log text file. Also doing the same on the terminal opens up a new terminal with nvim and the text file but if I do it through nnn it just opens the nvim process but not the terminal.

**Reply by charbelnicolas** [reply]
```
[Default Applications]

application/json=nvim.desktop;
application/pdf=org.pwmt.zathura.desktop;
application/octet-stream=nvim.desktop;
application/x-blender=blender.desktop;

audio/*=mpv.desktop;
image/*=feh.desktop;
video/*=mpv.desktop;
text/*=nvim.desktop;

x-scheme-handler/ftp=chromium.desktop;
x-scheme-handler/https=chromium.desktop;
x-scheme-handler/http=chromium.desktop;
x-scheme-handler/terminal=org.codeberg.dnkl.foot.desktop;

inode/directory=nnn.desktop;

[Added Associations]
```

**Reply by Anomalocaridid** [reply]
Thank you. Could you perhaps also post the file you are trying to open? Also, does this problem occur with other text files?

**Reply by charbelnicolas** [reply]
It occurs with all text files, anything that has to do with the terminal being open.

**Reply by charbelnicolas** [reply]
Blender files, pdf files, everything works as expected but as soon as neovim is involved, it won't open the foot terminal along with neovim, it will only spawn the nvim process.

**Reply by Anomalocaridid** [reply]
One thing I don't quite understand is how is `nnn` calling `handlr`? As far as I can tell, `nnn` does not use `handlr` in opening files by default at all.

Do you have the `xdg-utils-handlr` AUR package installed? Mainly asking because that's a possible explanation. I don't recommend installing it if it isn't already.

**Reply by charbelnicolas** [reply]
I actually had `xdg-utils-handlr` installed until today but it is no longer installed (I figured that out when trying to uninstall handlr and I got a dependency error). I do get the same behavior from nnn though. Nothing changed. I have `xdg-open` shadowed in `/.local/bin`.

**Reply by charbelnicolas** [reply]
I don't know how nnn handles things, it has just worked without any effort from my part.

**Reply by Anomalocaridid** [reply]
Could you remove the xdg-open shadowing script just to see what happens?

> I don't know how nnn handles things, it has just worked without any effort from my part.

I'm guessing it calls `xdg-open`, which was replaced by `xdg-utils-handlr` and is now shadowed by the script.

**Reply by charbelnicolas** [reply]
I did try it with and without the xdg-open shadowing, same outcome. Don't sweat it, I'll just keep using version 0.10.1 as long as it works. I do appreciate you taking the time for trying to fix it though, thank you and sorry for any rude behavior on my part.

**Reply by charbelnicolas** [reply]
BTW, this is what nnn says on its man page:

> nnn uses xdg-open (on Linux), [open(1)](https://man.archlinux.org/man/open.1.en) (on macOS), cygstart on (Cygwin) and open on (Haiku) as the desktop opener. It's also possible to specify a custom opener. See ENVIRONMENT section.

**Reply by Anomalocaridid** [reply]
> I did try it with and without the xdg-open shadowing, same outcome. Don't sweat it, I'll just keep using version 0.10.1 as long as it works.

Thank you for understanding. I'm sorry I could not figure out this problem. Once I've published it and it reaches the Arch repositories, it might still be worth trying the next release, v0.11.1, on the off chance a different fix also fixed your issue with nvim, foot, and nnn on accident.

> sorry for any rude behavior on my part.

No worries. I understand you were shocked and angered by one of your config files being randomly wiped and I'd say you more than made up for it by cooperating with me to fix your issues.

**Reply by charbelnicolas** [reply]
I'm just curious, what are your launch options for nnn and how do you open text files on nnn? Do you press the letter E to open it or do you press enter?

**Reply by charbelnicolas** [reply]
BTW, even when using version 0.10.1 when adding a labwc shortcut, this will open properly:
```
<keybind key="W-u">
	<action name="Execute">
		<command>xdg-open /home/charbel/Documents/design/blender/showroom.blend</command>
	</action>
</keybind>
```
![ss_2024_07_28_15_08_25](https://github.com/user-attachments/assets/387442bf-452c-4c15-b9e4-6e43aa4ef971)

But this will not:
```
<keybind key="W-u">
	<action name="Execute">
		<command>xdg-open /home/charbel/.log.txt</command>
	</action>
</keybind>
```
![ss_2024_07_28_15_06_28](https://github.com/user-attachments/assets/eb2c230c-1582-4df7-86c9-2689f5e11d98)

The foot terminal is missing...

**Reply by charbelnicolas** [reply]
I've fixed the issue with the following three commands, and honestly I didn't think it would be this easy:

`sudo pacman -Rs handlr-regex`, `rm /home/charbel/.local/bin/xdg-open` and then to finish it off, a simple `paru -S mimejs-git`.

So much for the blazingly fast rust implementation. It never ceases to amaze me how such a simple program like mimejs can do the neeeded work flawlessly without getting in the way.

A wise man once said it is always best to avoid the "It works on my computer™" developer mentality. How this ended up in the extra repository on arch linux is beyond me.

**Reply by Anomalocaridid** [reply]
> I'm just curious, what are your launch options for nnn and how do you open text files on nnn? Do you press the letter E to open it or do you press enter?

I don't normally use nnn. I was using the launch options you told me you were using, `-USJHd`. I was pressing enter and hitting the right arrow to open it.

**Reply by Anomalocaridid** [reply]
> I've fixed the issue with the following three commands, and honestly I didn't think it would be this easy:
> 
> `sudo pacman -Rs handlr-regex`, `rm /home/charbel/.local/bin/xdg-open` and then to finish it off, a simple `paru -S mimejs-git`.
> 
> So much for the blazingly fast rust implementation. It never ceases to amaze me how such a simple program like mimejs can do the neeeded work flawlessly without getting in the way.
> 
> A wise man once said it is always best to avoid the "It works on my computer™" developer mentality. How this ended up in the extra repository on arch linux is beyond me.

I am very, very sorry I couldn't figure out the problem preventing you from using `nvim`, I am glad that you have found an alternative that works for you, and I admit that yes, it is my fault and my fault alone it doesn't work any more.

**However, I would prefer that you do not go out of my way to insult me to my face.**

I would love nothing more than to fix your problem, but if I can't get the problem to occur on my own system, then how am I supposed to figure out how to fix it? **How am I supposed to know if any change I make even affects the problem at all?**

That is not "It works on my computer" mentality. That is me trying what I can and failing.

Of course, trying only gets so far, so by all means, use something that actually works for you. Trust me, I've been there. I get it. I couldn't care less if you or anyone else uses something else.

But don't take your frustrations out on me, the sole developer of this project, who maintains this project in my free time.

I hope someday I can fix this issue so others don't have to deal with it and I hope whoever's behind mimejs doesn't ever get a message like this from you.

Thank you for understanding.

**Reply by Celti** [reply]
> I hope someday I can fix this issue so others don't have to deal with it

I can reproduce this issue (or produce a similar one) *without* nnn by starting handlr without any stdin/stdout, e.g. `handlr open ~/test.txt 0<&- 1>&-`. Perhaps handlr isn't isolating child processes properly?

**Reply by Celti** [reply]
Aha, found it: https://github.com/Anomalocaridid/handlr-regex/blob/65797a7a9cd63b0e499fcd0d47992e8b32f5e615/src/common/desktop_entry.rs#L146

That should be `!config.terminal_output`.

**Reply by Anomalocaridid** [reply]
> I can reproduce this issue (or produce a similar one) _without_ nnn by starting handlr without any stdin/stdout, e.g. `handlr open ~/test.txt 0<&- 1>&-`. Perhaps handlr isn't isolating child processes properly?

I can reproduce this issue, although it does not look like it creates a persistent background process for me like the problem @charbelnicolas reported.

> Aha, found it:
> 
> https://github.com/Anomalocaridid/handlr-regex/blob/65797a7a9cd63b0e499fcd0d47992e8b32f5e615/src/common/desktop_entry.rs#L146
> 
> That should be `!config.terminal_output`.

Nice catch! I can confirm that fixes it! I'll definitely make sure to add a regression test as well.

**Reply by Anomalocaridid** [reply]
Ok, this should be solved in v0.11.2. I could not actually add a regression test due to some weirdness with how getting the `x-scheme-handler/terminal` program is currently implemented, but I intend to resolve it soon by reworking the implementation.

**Comment by charbelnicolas** [comment]
`application/json=nvim.desktop;` won't open in neovim either... This is inside neovim's `.desktop` file:

```
TryExec=nvim
Exec=nvim %F
Terminal=true
Type=Application
```

It does have `Terminal=true`

**Reply by charbelnicolas** [reply]
It never launches the foot terminal it just launches the nvim process without the actual terminal and you end up with something like this:

![ss_2024_07_25_14_22_46](https://github.com/user-attachments/assets/3783df78-7c54-43d6-8b34-01545fe2f003)

A ton of neovim processes lurking in the background...

**Comment by charbelnicolas** [comment]
Just downgraded to version 0.10.1 and everything is working as expected so I'll keep this version while a fix comes up for version 0.11.
